contracts/pcv/PCVGuardian.sol

```diff
@@ -76,114 +79,165 @@ contract PCVGuardian is IPCVGuardian, CoreRef {
     function unsetSafeAddresses(address[] calldata _safeAddresses)
         external
         override
-        hasAnyOfThreeRoles(TribeRoles.GOVERNOR, TribeRoles.GUARDIAN, TribeRoles.PCV_GUARDIAN_ADMIN)
+        hasAnyOfThreeRoles(TribeRoles.PCV_GUARDIAN_ADMIN, TribeRoles.GUARDIAN, TribeRoles.GOVERNOR)
     {
-        require(_safeAddresses.length != 0, "empty");
+        require(_safeAddresses.length != 0, "PCVGuardian: empty");
         for (uint256 i = 0; i < _safeAddresses.length; i++) {
             _unsetSafeAddress(_safeAddresses[i]);
         }
     }
 
-    /// @notice governor-or-guardian-only method to withdraw funds from a pcv deposit, by calling the withdraw() method on it
-    /// @param pcvDeposit the address of the pcv deposit contract
-    /// @param safeAddress the destination address to withdraw to
-    /// @param amount the amount to withdraw
-    /// @param pauseAfter if true, the pcv contract will be paused after the withdraw
-    /// @param depositAfter if true, attempts to deposit to the target PCV deposit
-    function withdrawToSafeAddress(
+    /// @notice modifier to factorize the logic in all withdrawals :
+    /// - first, ensure the deposit to withdraw from is unpaused
+    /// - second, perform withdrawal
+    /// - third, re-pause deposit if it was paused
+    /// - finally, call pcvDeposit.deposit() if depositAfter = true
+    modifier beforeAndAfterWithdrawal(
         address pcvDeposit,
         address safeAddress,
-        uint256 amount,
-        bool pauseAfter,
         bool depositAfter
-    ) external override hasAnyOfThreeRoles(TribeRoles.GOVERNOR, TribeRoles.PCV_SAFE_MOVER_ROLE, TribeRoles.GUARDIAN) {
-        require(isSafeAddress(safeAddress), "Provided address is not a safe address!");
+    ) {
+        {
+            // scoped in this modifier to prevent stack to deep errors & enforce consistent acl
+            ICore _core = core();
+            require(
+                _core.hasRole(TribeRoles.GUARDIAN, msg.sender) ||
+                    _core.hasRole(TribeRoles.PCV_SAFE_MOVER_ROLE, msg.sender) ||
+                    _core.hasRole(TribeRoles.GOVERNOR, msg.sender),
+                "UNAUTHORIZED"
+            );
+            require(isSafeAddress(safeAddress), "PCVGuardian: address not whitelisted");
+        }
 
-        pcvDeposit._ensureUnpaused();
+        bool paused = pcvDeposit._paused();
+        if (paused) {
```

Comment:
Why not just use ensureUnpaused here?

Reply:
`ensureUnpaused` reads the `pcvDeposit.paused()` but does not return it. We need to read the paused state before withdrawal, because after withdrawal we want to re-pause if the deposit was paused in the first place